File: 22. Generate Parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList();
        boringStuff(ans, "", 0, 0, n);
        return ans;
    }

    public void boringStuff(List<String> ans, String cur, int open, int close, int n){
        if (cur.length() == n * 2) {
            ans.add(cur);
            return;
        }

        if (open < n)
            boringStuff(ans, cur+"(", open+1, close, n);
        if (close < open)
            boringStuff(ans, cur+")", open, close+1, n);
    }
}

File: 22. Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ret;
        if(n<=0){
            ret.push_back("");
            return ret;
        }
        string s="(";
        dfs(ret, 1, 0, n, s);
        return ret;
    }
     void dfs(vector<string> &ret, int l, int r, int n, string &s){
         // cout<<l<<r<<n<<endl;
         if(l==r&&r==n){
             ret.push_back(s);
             return;
         }
         if(l>n||r>n||r>l) return;
         
         s+='(';
         dfs(ret, l+1, r, n, s);
         s.pop_back();
         s+=')';
         dfs(ret, l, r+1, n, s);
         s.pop_back();
         
     }
};